File: hilti/python/hilti/codegen/overlay.py
# $Id$
#
# Code generator for overlay instructions.
#
# Overlays keep two kinds of state:
#
#   - The starting position corresponding to offset 0. This is set by the
#     "attach" instruction.
#
#   - For each field that is a dependency for determining another field's
#     starting offset, we cache its end position once we have calculated it
#     once. 


import llvm.core

from hilti.core import *
from hilti import instructions
from codegen import codegen

import sys

_doc_c_conversion = _doc_c_conversion = """An ``overlay`` is mapped to a ``hlt_overlay``."""

def _iterType():
    return codegen.llvmTypeConvert(type.IteratorBytes(type.Bytes()))

def _arraySize(overlay):
    return 1 + overlay.numDependencies()

def _llvmOverlayType(overlay):
    return llvm.core.Type.array(_iterType(), _arraySize(overlay))

def _unsetIter():
    return codegen.llvmConstDefaultValue(type.IteratorBytes(type.Bytes()))

@codegen.typeInfo(type.Overlay)
def _(t):
    typeinfo = codegen.TypeInfo(t)
    typeinfo.c_prototype = "hlt_overlay";
    return typeinfo

@codegen.llvmDefaultValue(type.Overlay)
def _(t):
    return llvm.core.Constant.array(_iterType(), [_unsetIter()] * _arraySize(t))

@codegen.llvmType(type.Overlay)
def _(t, refine_to):
    return _llvmOverlayType(t)

@codegen.when(instructions.overlay.Attach)
def _(self, i):
    overlayt = i.op1().type()
    
    # Save the starting offset.
    ov = self.llvmOp(i.op1())
    iter = self.llvmOp(i.op2())
    ov = codegen.llvmInsertValue(ov, 0, iter)
    
    # Clear the cache
    for j in range(1, _arraySize(overlayt) + 1):
        ov = codegen.llvmInsertValue(ov, j, _unsetIter())
        
    # Need to rewrite back into the original value.
    codegen.llvmStoreInTarget(i.op1(), ov)

def _isNull(ov, idx, block_yes, block_no):    
    # FIXME: I guess we should do this check somewhere in the bytes code as
    # technically we can't know here what the internal representation of an
    # iterator is ...
    val = codegen.llvmExtractValue(ov, idx)
    chunk = codegen.llvmExtractValue(val, 0)
    cmp = codegen.builder().icmp(llvm.core.IPRED_EQ, chunk, llvm.core.Constant.null(codegen.llvmTypeGenericPointer()))
    codegen.builder().cbranch(cmp, block_yes, block_no)
    
@codegen.when(instructions.overlay.Get)
def _(self, i):
    overlayt = i.op1().type()
    ov = self.llvmOp(i.op1())
    offset0 = codegen.llvmExtractValue(ov, 0)
    
    # Generate the unpacking code for one field, assuming all dependencies are
    # already resolved.
    def _emitOne(ov, field):
        offset = field.offset()
        
        if offset == 0:
            # Starts right at the beginning.
            begin = offset0
            
        elif offset > 0:
            # Static offset. We can calculate the starting position directly.
            begin = self.llvmGenerateCCallByName("hlt::bytes_pos_incr_by", 
            [offset0, codegen.llvmConstInt(offset, 32)], [type.IteratorBytes(type.Bytes()), type.Integer(32)], llvm_args=True)
            
        else:
            # Offset is not static but our immediate dependency must have
            # already been calculated at this point so we take it's end
            # position out of the cache.
            deps = field.dependants()
            assert deps
            dep = overlayt.field(deps[-1]).depIndex()
            assert dep > 0
            begin = codegen.llvmExtractValue(ov, dep)
                    
        arg = self.llvmOp(field.arg()) if field.arg() else None
        (val, end) = codegen.llvmUnpack(field.type(), begin, None, field.fmt(), arg)
        
        if field.depIndex() >= 0:
            ov = codegen.llvmInsertValue(ov, field.depIndex(), end)
        
        return (ov, val)

    # Generate code for all depedencies.
    def _makeDep(ov, field):

        if not field.dependants():
            return ov

        # Need a tmp or a PHI node. LLVM's mem2reg should optimize the tmp
        # away so this is easier. 
        # FIXME: Check that mem2reg does indeed.
        new_ov = codegen.llvmAlloca(ov.type)         
        dep = overlayt.field(field.dependants()[-1])
        block_cont = self.llvmNewBlock("have-%s" % dep.name)
        block_cached = self.llvmNewBlock("cached-%s" % dep.name)
        block_notcached = self.llvmNewBlock("not-cached-%s" % dep.name)

        _isNull(ov, dep.depIndex(), block_notcached, block_cont)
        
        self.pushBuilder(block_notcached)
        ov = _makeDep(ov, dep)
        self.builder().store(ov, new_ov)
        self.builder().branch(block_cont)
        self.popBuilder()
        
        self.pushBuilder(block_cached)
        self.builder().store(ov, new_ov)
        self.builder().branch(block_cont)
        self.popBuilder()
        
        self.pushBuilder(block_cont)
        ov = self.builder().load(new_ov)
        (ov, val) = _emitOne(ov, dep)
        return ov
    
    # Make sure we're attached.
    block_attached = self.llvmNewBlock("attached")
    block_notattached = self.llvmNewBlock("not-attached")
    
    self.pushBuilder(block_notattached)
    self.llvmRaiseExceptionByName("hlt_exception_overlay_not_attached", i.location())
    self.popBuilder()

    _isNull(ov, 0, block_notattached, block_attached)
    
    self.pushBuilder(block_attached) # Leave on stack.
    
    field = overlayt.field(i.op2().value())
    ov = _makeDep(ov, field)
    (ov, val) = _emitOne(ov, field)
    self.llvmStoreInTarget(i.target(), val)

        
    

File: hilti/codegen/instructions/old/overlay.cc

#include <hilti.h>

#include "../stmt-builder.h"

using namespace hilti;
using namespace codegen;


void StatementBuilder::visit(statement::instruction::overlay::Attach* i)
{
    /*
        auto op1 = cg()->llvmValue(i->op1(), X);
        auto op2 = cg()->llvmValue(i->op2(), X);
    */

    /*
        CodeGen::expr_list args;
        args.push_back(i->op1());
        args.push_back(i->op2());
        cg()->llvmCall("hlt::X", args);
    */


    /*
        def _codegen(self, cg):
            overlayt = self.op1().type()

            # Save the starting offset.
            ov = cg.llvmOp(self.op1())
            iter = cg.llvmOp(self.op2())
            ov = cg.llvmInsertValue(ov, 0, iter)

            # Clear the cache
            end = bytes.llvmEnd(cg)
            for j in range(1, overlayt._arraySize()):
                ov = cg.llvmInsertValue(ov, j, end)

            # Need to rewrite back into the original value.
            self.op1().llvmStore(cg, ov)

    */
}

void StatementBuilder::visit(statement::instruction::overlay::Get* i)
{
    /*
        auto op1 = cg()->llvmValue(i->op1(), X);
        auto op2 = cg()->llvmValue(i->op2(), X);

        auto result = builder()->

        cg()->llvmStore(i, result);
    */

    /*
        CodeGen::expr_list args;
        args.push_back(i->op1());
        args.push_back(i->op2());

        auto result = cg()->llvmCall("hlt::X", args);

        cg()->llvmStore(i, result);
    */


    /*
        def _codegen(self, cg):
            overlayt = self.op1().type()
            ov = cg.llvmOp(self.op1())
            offset0 = cg.llvmExtractValue(ov, 0)

            # Generate the unpacking code for one field, assuming all dependencies are
            # already resolved.
            def _emitOne(ov, field):
                offset = field.offset()

                if offset == 0:
                    # Starts right at the beginning.
                    begin = offset0

                elif offset > 0:
                    # Static offset. We can calculate the starting position directly.
                    arg1 = operand.LLVM(offset0, type.IteratorBytes())
                    arg2 = operand.Constant(constant.Constant(offset, type.Integer(32)))
                    begin = cg.llvmCallC("hlt::bytes_pos_incr_by", [arg1, arg2])

                else:
                    # Offset is not static but our immediate dependency must have
                    # already been calculated at this point so we take it's end
                    # position out of the cache.
                    deps = field.dependants()
                    assert deps
                    dep = overlayt.field(deps[-1]).depIndex()
                    assert dep > 0
                    begin = cg.llvmExtractValue(ov, dep)

                if field.arg():
                    arg = cg.llvmOp(field.arg())
                    arg = operand.LLVM(arg, field.arg().type())
                else:
                    arg = None

                (val, end) = cg.llvmUnpack(field.type(), begin, None, field.fmt(), arg)

                if field.depIndex() >= 0:
                    ov = cg.llvmInsertValue(ov, field.depIndex(), end)

                return (ov, val)

            # Generate code for all depedencies.
            def _makeDep(ov, field):

                if not field.dependants():
                    return ov

                # Need a tmp or a PHI node. LLVM's mem2reg should optimize the tmp
                # away so this is easier.
                # FIXME: Check that mem2reg does indeed.
                new_ov = cg.llvmAlloca(ov.type)
                dep = overlayt.field(field.dependants()[-1])
                block_cont = cg.llvmNewBlock("have-%s" % dep.name)
                block_cached = cg.llvmNewBlock("cached-%s" % dep.name)
                block_notcached = cg.llvmNewBlock("not-cached-%s" % dep.name)

                _isEnd(cg, ov, dep.depIndex(), block_notcached, block_cont)

                cg.pushBuilder(block_notcached)
                ov = _makeDep(ov, dep)
                cg.builder().store(ov, new_ov)
                cg.builder().branch(block_cont)
                cg.popBuilder()

                cg.pushBuilder(block_cached)
                cg.builder().store(ov, new_ov)
                cg.builder().branch(block_cont)
                cg.popBuilder()

                cg.pushBuilder(block_cont)
                ov = cg.builder().load(new_ov)
                (ov, val) = _emitOne(ov, dep)
                return ov

            # Make sure we're attached.
            block_attached = cg.llvmNewBlock("attached")
            block_notattached = cg.llvmNewBlock("not-attached")

            cg.pushBuilder(block_notattached)
            cg.llvmRaiseExceptionByName("hlt_exception_overlay_not_attached", self.location())
            cg.popBuilder()

            _isEnd(cg, ov, 0, block_notattached, block_attached)

            cg.pushBuilder(block_attached) # Leave on stack.

            field = overlayt.field(self.op2().value().value())
            ov = _makeDep(ov, field)
            (ov, val) = _emitOne(ov, field)
            cg.llvmStoreInTarget(self, val)

    */
}
